File: src/main/java/com/company/erp/features/employees/dto/EmployeeDto.java
package com.company.erp.features.employees.dto;

import com.company.erp.features.employees.entity.Employee;
import com.company.erp.features.employees.entity.EmployeeStatus;

import java.time.Instant;
import java.time.LocalDate;
import java.time.Period;
import java.time.ZoneId;

public record EmployeeDto(
        Long id,
        Long userId,
        String employeeNo,
        String fullName,
        String workEmail,
        String phone,
        String position,
        String department,
        LocalDate hireDate,
        LocalDate dateOfBirth,
        String gender,
        String address,
        String avatarUrl,
        String avatarContentType,
        Instant avatarUploadedAt,
        String emergencyContact,
        String emergencyPhone,
        Instant lastLoginAt,
        /** Derived from hireDate at read time, formatted like "2y 5m". null if no hireDate. */
        String tenure,
        EmployeeStatus status
) {
    public static EmployeeDto from(Employee e) {
        return new EmployeeDto(
                e.getId(),
                e.getUserId(),
                e.getEmployeeNo(),
                e.getFullName(),
                e.getWorkEmail(),
                e.getPhone(),
                e.getPosition(),
                e.getDepartment(),
                e.getHireDate(),
                e.getDateOfBirth(),
                e.getGender(),
                e.getAddress(),
                e.getAvatarUrl(),
                e.getAvatarContentType(),
                e.getAvatarUploadedAt(),
                e.getEmergencyContact(),
                e.getEmergencyPhone(),
                e.getLastLoginAt(),
                formatTenure(e.getHireDate()),
                e.getStatus());
    }

    private static String formatTenure(LocalDate hireDate) {
        if (hireDate == null) return null;
        LocalDate today = LocalDate.now(ZoneId.of("UTC"));
        if (hireDate.isAfter(today)) return "0y 0m";
        Period p = Period.between(hireDate, today);
        return p.getYears() + "y " + p.getMonths() + "m";
    }
}

File: app/Features/Employees/Dto/EmployeeDto.php
<?php

namespace App\Features\Employees\Dto;

use App\Features\Employees\Models\Employee;
use Carbon\CarbonImmutable;

/**
 * Port of the Spring EmployeeDto record, including the derived `tenure`
 * ("2y 5m") computed from hire_date at read time.
 */
final class EmployeeDto
{
    /**
     * @return array<string, mixed>
     */
    public static function from(Employee $e): array
    {
        return [
            'id' => (int) $e->id,
            'userId' => $e->user_id !== null ? (int) $e->user_id : null,
            'employeeNo' => $e->employee_no,
            'fullName' => $e->full_name,
            'workEmail' => $e->work_email,
            'phone' => $e->phone,
            'position' => $e->position,
            'department' => $e->department,
            'hireDate' => $e->hire_date ? $e->hire_date->format('Y-m-d') : null,
            'dateOfBirth' => $e->date_of_birth ? $e->date_of_birth->format('Y-m-d') : null,
            'gender' => $e->gender,
            'address' => $e->address,
            'avatarUrl' => $e->avatar_url,
            'avatarContentType' => $e->avatar_content_type,
            'avatarUploadedAt' => $e->avatar_uploaded_at ? $e->avatar_uploaded_at->utc()->format('Y-m-d\TH:i:s\Z') : null,
            'emergencyContact' => $e->emergency_contact,
            'emergencyPhone' => $e->emergency_phone,
            'lastLoginAt' => $e->last_login_at ? $e->last_login_at->utc()->format('Y-m-d\TH:i:s\Z') : null,
            'tenure' => self::formatTenure($e->hire_date),
            'status' => $e->status,
        ];
    }

    private static function formatTenure($hireDate): ?string
    {
        if ($hireDate === null) {
            return null;
        }
        $hire = CarbonImmutable::parse($hireDate->format('Y-m-d'), 'UTC')->startOfDay();
        $today = CarbonImmutable::now('UTC')->startOfDay();
        if ($hire->greaterThan($today)) {
            return '0y 0m';
        }
        $diff = $hire->diff($today);
        return $diff->y.'y '.$diff->m.'m';
    }
}
